### ISSUE TITLE
"shader_get_param_list" returns texture parameters last

### ISSUE THREAD

**Issue by Arnklit** [godotengine/godot#45329]
**Godot version:**
3.2.3.Stable

**OS/device including version:**
NA

**Issue description:**

"VisualServer.shader_get_param_list()" returns texture parameters last. This means that ShaderMaterials display their uniforms in the inspector in a different order than they are declared in the shader. This makes it difficult to make nice custom inspectors for shader parameters and is confusing when using a custom shader in general.

![image](https://user-images.githubusercontent.com/4955051/105184461-45554380-5b27-11eb-9f58-83dcdaf49844.png)

![image](https://user-images.githubusercontent.com/4955051/105184512-543bf600-5b27-11eb-9acc-61d2aec92c87.png)

Result of printing "shader_get_param_list"
```
[ {
"class_name": "",
"hint": 0,
"hint_string": "",
"name": "float_uniform",
"type": 3,
"usage": 7
}, {
"class_name": "",
"hint": 0,
"hint_string": "",
"name": "int_uniform",
"type": 2,
"usage": 7
}, {
"class_name": "",
"hint": 0,
"hint_string": "",
"name": "vec_uniform",
"type": 7,
"usage": 7
}, {
"class_name": "",
"hint": 17,
"hint_string": "Texture",
"name": "sampler_uniform1",
"type": 17,
"usage": 7
}, {
"class_name": "",
"hint": 17,
"hint_string": "Texture",
"name": "sampler_uniform2",
"type": 17,
"usage": 7
}, {
"class_name": "",
"hint": 17,
"hint_string": "Texture",
"name": "sampler_uniform3",
"type": 17,
"usage": 7
} ]
```

**Steps to reproduce:**
Open the attached project and observe the shader on the object compared to it's shader parameter inspector. Use the "print params" checkbox to see the shader_get_param_list output.

**Minimal reproduction project:**

[shader_get_param_list.zip](https://github.com/godotengine/godot/files/5842866/shader_get_param_list.zip)

**Comment by Calinou** [comment]
I agree this should be changed to match the declaration order. Feel free to open a pull request for this :slightly_smiling_face:

**Comment by Arnklit** [comment]
I'll try and see if I can figure out what's causing it and do the pull request. Seems like a good first bug fix for me to try.

**Comment by Arnklit** [comment]
I thought https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/renderer_scene_render_forward.cpp line 335 looked like a likely candidate to where this issues is
```
void RendererSceneRenderForward::ShaderData::get_param_list(List<PropertyInfo> *p_param_list) const {
	Map<int, StringName> order;

	for (Map<StringName, ShaderLanguage::ShaderNode::Uniform>::Element *E = uniforms.front(); E; E = E->next()) {
		if (E->get().scope != ShaderLanguage::ShaderNode::Uniform::SCOPE_LOCAL) {
			continue;
		}

		if (E->get().texture_order >= 0) {
			order[E->get().texture_order + 100000] = E->key();
		} else {
			order[E->get().order] = E->key();
		}
	}

	for (Map<int, StringName>::Element *E = order.front(); E; E = E->next()) {
		PropertyInfo pi = ShaderLanguage::uniform_to_property_info(uniforms[E->get()]);
		pi.name = E->get();
		p_param_list->push_back(pi);
	}
}
```
Adding 10000 to the texture_order is to make sure they are at the end I assume, I don't really get what's going on here.

I tried commenting out the if statement and always doing the or part and that causes the texture slots to always be in front and override each other as well it seemed, so obviously this was done to work around some issue.

Someone a bit more familiar with this might need to chime in.
